### DISCUSSION TITLE
An all-in-one command like traverse, but that restacks PRs on Github?

### DISCUSSION THREAD

**Discussion by chriscz** [VirtusLab/git-machete#1214] [Q&A]
Git Machete newb here. I've been evaluating different git tools for stacked PRs and I'm currently trailing out machete. 

**My Question**

Is there some command that would do the following:
1. Fetch all the root branches 
2. Traverse, find new fork point, rebase all my feature branches (those in the `edit` screen)
3. Push up to origin these branches
4. Restack or Retarget all the branches that have PRs associated with them

Or, is there a way I could do something like

`git machete branches traverse --pull --push pr --retarget`

**General Experience and Suggestions so far**

My experience so far has been pleasant the the docs have been very clear and a real help in explaining both _what_ and _why_. Machete has been quite useful to me, primarily the traverse command and the basic GH integration (so I don't need to us `gh` all the time)

However, I do feel that the CLI could benefit from a more resource-oriented design with nouns and verbs, 
and possibly have a limited set of context-dependent convenience methods available at the top-level working on the current branch or pr for example:

```
git machete branches traverse
git machete annos pull
git machete pr create (whether for github, bitbucket, gitlab or gittea for example)
```

**Comment by PawelLipski** [comment]
Hey there,

If it comes to auto-restacking of branches during `git machete traverse`... yes, I've got the same idea recently ;D lemme open an issue for that: #1215. I'm wrapping up GitLab integration, this will be one of the first things I'll do once GitLab's done.

As for resource-oriented CLI... as you can see from the docs, the CLI grew more than I initially expected, thanks to various user requests and contributions 😅 indeed, a major revamp could make the CLI clearer and more predictable... but I need to weigh in the habits of the long-term users (as this software exists since early 2018, and gained significant popularity around mid-2022).

The alternative could be to add the resources like `branches`, `annos`, `prs` as new commands in the CLI, without removing the existing commands... but then the CLI would expand significantly and might end up too confusing for the newcomers :/

**Reply by chriscz** [reply]
Ah that's great. I though I was just missing something obvious. 

It's probably a good idea to somehow keep anyone that's on an old version to continue using the previous CLI, as you rightly note that 'em fingers get used to it :wink:.

I also wonder how powerful it could be to have a text-based protocol where you could run `git machete t -i` (for interactive mode), and then specify actions on each branch? I'm not certain of it's usefulness just yet, but I'm quite inspired by git's interactive rebase!

I hope to give `git-branchless` a go at some point too, as I'm currently in the dating phase of exploring stacked commits :)

PS. Do you have any recommended reading for such a workflow, or things you wish you had read sooner than later? @PawelLipski

**Reply by PawelLipski** [reply]
> I also wonder how powerful it could be to have a text-based protocol where you could run git machete t -i (for interactive mode), and then specify actions on each branch? I'm not certain of it's usefulness just yet, but I'm quite inspired by git's interactive rebase!

Hmm the only thing that's somewhat similar for now are _branch qualifiers_:
* `push=no`
* `rebase=no`
* `slide-out=no`
* `update=merge`

The first three tell `git machete traverse` to skip certain operation (like, don't push someone else's branch). The last one makes traverse sync branch to its parent via merge and not rebase (but careful with that, as [merges don't work well with stacked branches](https://github.com/VirtusLab/git-machete/?tab=readme-ov-file#can-i-use-git-merge-for-dealing-with-stacked-prs)). 

We've just observed that for the given branch, it's usually the same set of operations that you'd like to perform. So **no** push/rebase on someone else's branches, and **no** sliding out of `develop` branch that's based on `master`.

Obviously, there's also fully interactive mode &mdash; if you don't specify `--yes` to `traverse`, it'll ask you to confirm each operation.

Soo... together, this options covered almost all useful flavors of (non)interactivity in `traverse`, from our experience :)

**Reply by PawelLipski** [reply]
> PS. Do you have any recommended reading for such a workflow, or things you wish you had read sooner than later? @PawelLipski

Huh good question... the problem is that we've started with `git-machete` at a point when there was barely any similar tool around 🤔 so we've mostly figured it out based on our own use + obviously feature requests from users. TBH I can't recommend much more than is already in [Reference](https://github.com/VirtusLab/git-machete?tab=readme-ov-file#reference)... but if you find a good read that goes beyond the scope of these, pls share!

**Comment by PawelLipski** [comment]
@chriscz this took me a while, but partial GitHub/GitLab support in `git machete traverse` (via flags like `--sync-github-prs`) is almost ready on PR #1340, to be released in v3.32.0 📦 `traverse` will then suggest retargeting branches that have a different parent in machete than in GitHub PR

**Comment by PawelLipski** [comment]
Long done at this point, see `git machete traverse --sync-github-prs`
